src/parameterizations/lateral/MOM_lateral_mixing_coeffs.F90

```diff
@@ -1128,6 +1153,123 @@ subroutine calc_slope_functions_using_just_e(h, G, GV, US, CS, e)
+         "UH_grad is not associated with use_gradient_model.")
+  if (.not. allocated(CS%VH_grad)) call MOM_error(FATAL, "calc_slope_function:"// &
+         "VH_grad is not associated with use_gradient_model.")
+
+  is = G%isc ; ie = G%iec ; js = G%jsc ; je = G%jec ; nz = GV%ke
+
+  h_neglect = GV%H_subroundoff
+  H_cutoff = real(2*nz) * (GV%Angstrom_H + h_neglect)
+
+
+  ! Set the length scale at u-points for the gradient model
+  do j=js,je ; do I=is-1,ie
+    Lgrid = sqrt(G%dxCu(I,j)**2 + G%dyCu(I,j)**2)
+    CS%L2grad_u(I,j) = 1.0 * Lgrid**2
+  enddo ; enddo
+  ! Set length scale at v-points for the gradient model
+  do J=js-1,je ; do i=is,ie
+    Lgrid = sqrt(G%dxCv(i,J)**2 + G%dyCv(i,J)**2)
+    CS%L2grad_v(i,J) = 1.0 * Lgrid**2
+  enddo ; enddo
+
+  do k=nz,CS%VarMix_Ktop,-1
+
+    ! Calculate the gradient slopes Ux_Hx, Vx_Hx, Uy_Hy, Vy_Hy on u- and v-points respectively
+    do j=js-1,je+1 ; do I=is-1,ie
+      Ux_Hx(I,j) = ((G%IdxCu(I+1,j)*G%IdyCu(I+1,j)*uh(I+1,j,K)) - (G%IdxCu(I,j)*G%IdyCu(I,j)*uh(I,j,k))) * ( &
+                   G%IareaT(I+1,j) + G%IareaT(I,j)) * G%dyT(I,j) * ((h(i+1,j,k) - h(i,j,k))/( &
+                   h(i+1,j,k) + h(i,j,k) + h_neglect))
+      Vx_Hx(I,j) = ((G%IdxCv(I+1,j)*G%IdxCv(I+1,j)*vh(I+1,j,K)) - (G%IdxCv(I,j)*G%IdxCv(I,j)*vh(I,j,k))) * ( &
+                   G%IareaT(I+1,j) + G%IareaT(I,j)) * G%dyT(I,j) * ((h(i+1,j,k) - h(i,j,k))/( &
+                   h(i+1,j,k) + h(i,j,k) + h_neglect))
+    ! Mask slopes where interface intersects topography
+      if (min(h(i,j,k),h(i+1,j,k)) < H_cutoff) Ux_Hx(I,j) = 0.
+      if (min(h(i,j,k),h(i+1,j,k)) < H_cutoff) Vx_Hx(I,j) = 0.
+    enddo ; enddo
+    do J=js-1,je ; do i=is-1,ie+1
+      Uy_Hy(i,J) = ((G%IdyCu(i,J+1)*G%IdyCu(i,J+1)*uh(i,J+1,K)) - (G%IdyCu(i,J)*G%IdyCu(i,J)*uh(i,J,k))) * ( &
+                   G%IareaT(i,J+1) + G%IareaT(i,J)) * G%dxT(i,J) * ((h(i,j+1,k) - h(i,j,k))/( &
+                   h(i,j+1,k) + h(i,j,k) + h_neglect))
+      Vy_Hy(i,J) = ((G%IdyCv(i,J+1)*G%IdxCv(i,J+1)*vh(i,J+1,K)) - (G%IdyCv(i,J)*G%IdxCv(i,J)*vh(i,J,k))) * ( &
+                   G%IareaT(i,J+1) + G%IareaT(i,J)) * G%dxT(I,j) * ((h(i,j+1,k) - h(i,j,k))/( &
+                   h(i,j+1,k) + h(i,j,k) + h_neglect))
+    ! Mask slopes where interface intersects topography
+      if (min(h(i,j,k),h(i,j+1,k)) < H_cutoff) Uy_Hy(I,j) = 0.
+      if (min(h(i,j,k),h(i,j+1,k)) < H_cutoff) Vy_Hy(I,j) = 0.
+    enddo ; enddo
+
+    do j=js,je ; do I=is-1,ie
+      graduh = Ux_Hx(I,j) + 0.25 * ( (Uy_Hy(I,j) + Uy_Hy(I+1,j-1)) + (Uy_Hy(I,j-1) + Uy_Hy(I+1,j)) )
+      UH_grad_local(I,j,k) = graduh
+    enddo ; enddo
+
+    do J=js-1,je ; do i=is,ie
+      gradvh = 0.25 * ( (Vx_Hx(i,J) + Vx_Hx(i-1,J+1)) + (Vx_Hx(i-1,J) + Vx_Hx(i,J+1)) ) + Vy_Hy(i,J)
+      VH_grad_local(i,J,k) = gradvh
+    enddo ; enddo
+
+  enddo ! k
+
+  !$OMP parallel do default(shared)
+  do j=js,je
+
+    do k=nz,CS%VarMix_Ktop,-1 ; do I=is-1,ie
+      CS%UH_grad(I,j,k) = UH_grad_local(I,j,k)
```

Comment:
I suspect that `CS%UH_grad` could be set directly where `graduh` is being set about 15 lines above, which would save an unnecessary copy of a 3-d array.